Add Maximum Subarray Sum Solution (#947)
cpinitiative/usaco-guide: content/3_Silver/Prefix_Sums_2.mdx

---
id: prefix-sums-2
title: 'More on Prefix Sums'

> author: Darren Yao, Eric Wei, Neo Wang, Jesse Choe

Reviewer: is this (removing Qi Wang) intentional?

Author: Oops no..

Author: I didn't even notice that :cry:

<FocusProblem problem={problems.maxsum} />

### Solution

<Spoiler title="Solution">

> Consider the prefix sum array $p$. The subarray sum $a_j \dots a_{i-1}$ where $j\leq i$ is $p[i]-p[j]$, so we are looking for the maximum possible value of $p[i]-p[j]$ over $0 \leq i \leq j \leq N$. 

Reviewer 2: use \texttt for variables

Reviewer 2: not one letter vars tho

Author: Can you commit a suggestion, so that ik what you're talking ab??

Reviewer 3: i think the `p[i]` and `p[j]`, right?

Reviewer 2: yes use \texttt{p}[i] etc

Reviewer 3: I think `\texttt{p[i]}` looks better? idk
![image](https://user-images.githubusercontent.com/63619830/111053264-339d6780-8417-11eb-9479-83f3c27c2175.png)

Reviewer 2: hmm i like the 2nd one but the 2nd one is standard anyways

For a fixed right bound $i$, the maximum subarray sum is

> $$$

Reviewer 3: Why are there 3 `$`? Shouldn't 2 work?

Author: idk i just commited what the other ppl recommended.

> $$$
> p[i]-\min_{j\leq i}{p[j]}
> $$$

Reviewer 3: ```suggestion
$$
p[i]-\min_{j\leq i}{p[j]}
$$
```

> Thus, we can keep a running minimum to store $\min_{j\leq i}{p[j]}$ as we iterate through $i$. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.

Reviewer 3: ```suggestion
Thus, we can keep a running minimum to store $\min\limits_{j\leq i}{p[j]}$ as we iterate through $i$. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.
```
move the j \leq i under \min

<LanguageSection>

<CPPSection>

C++ code:

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;

> 	vl dp(n+1, 0); // Prefix sum array initially filled with 0's

Reviewer 4: ```suggestion
	vl pfx(n+1, 0); // Prefix sum array initially filled with 0's
```

	for(int i=1; i<=n; i++){
		ll x; cin >> x;

> 		dp[i] = dp[i-1] + x; // Compute the prefix sum at each element

Reviewer 4: ```suggestion
		pfx[i] = pfx[i-1] + x; // Compute the prefix sum at each element
```

	}

> 	ll max_subarray_sum = dp[0], min_prefix_sum = dp[0];

Reviewer 4: ```suggestion
	ll max_subarray_sum = pfx[0], min_prefix_sum = pfx[0];
```

	for(int i=1; i<=n; i++){
		// Max subarray sum is the maximum difference between two prefix sums

> 		max_subarray_sum = max(max_subarray_sum, dp[i] - min_prefix_sum);

Reviewer 4: ```suggestion
		max_subarray_sum = max(max_subarray_sum, pfx[i] - min_prefix_sum);
```

> 		min_prefix_sum = min(min_prefix_sum, dp[i]);

Reviewer 4: ```suggestion
		min_prefix_sum = min(min_prefix_sum, pfx[i]);
```

	}
	cout << max_subarray_sum << endl;
}
```

</CPPSection>

</LanguageSection>

</Spoiler>

<Warning>

This problem has a solution known as [Kadane's Algorithm](https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm). Please don't use that solution; try to solve it with prefix sums.

</Warning>

<Spoiler title = "Alternative Solution: Kadane's Algorithm">

Kadane's Algorithm finds the largest subarray sum using a greedy method. Check the video below for details:

[Video explanation of the max subarray sum using Kadane's Algorithm](https://www.youtube.com/watch?v=jnoVtCKECmQ&t=611s).

<LanguageSection>

<CPPSection>

## C++ Implementation

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl x(n);
	for(int i=0; i<n; i++){
		cin >> x[i];
	}
	ll current_sum = x[0], max_subarray_sum = x[0];
	for(int i=1; i<n; i++){
		current_sum = max(current_sum + x[i], x[i]); 

> 		// Continue the subarray sum or start a new subarray sum beginning at the current element

Reviewer 3: This line is too long, maybe put a line break before "or"?

Author: ok.

Reviewer 3: This hasn't been fixed yet...

Author: Oops - forgot.

For a fixed right bound $j$, the maximum subarray sum is

$$
p[j]-\min_{i\leq j}{p[i]}
$$

Thus, we can keep a running minimum to store $\min\limits_{i\leq j}{p[i]}$ as we iterate through $i$. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.

<LanguageSection>

<CPPSection>

## C++ Implementation

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl pfx(n+1, 0); // Prefix sum array initially filled with 0's
	for(int i=1; i<=n; i++){
		ll x; cin >> x;
		pfx[i] = pfx[i-1] + x; // Compute the prefix sum at each element
	}
	ll max_subarray_sum = pfx[0], min_prefix_sum = pfx[0];
	for(int i=1; i<=n; i++){
		// Max subarray sum is the maximum difference between two prefix sums
		max_subarray_sum = max(max_subarray_sum, pfx[i] - min_prefix_sum);
		min_prefix_sum = min(min_prefix_sum, pfx[i]);
	}
	cout << max_subarray_sum << endl;
}
```

</CPPSection>

</LanguageSection>

</Spoiler>

> <Warning>

Reviewer 2: why is this a warning?

Author: I thought it was better as a warning...

This problem has a solution known as [Kadane's Algorithm](https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm). Please don't use that solution; try to solve it with prefix sums.

</Warning>

<Spoiler title = "Alternative Solution: Kadane's Algorithm">

Kadane's Algorithm finds the largest subarray sum using a greedy method. Check this

> [video tutorial](https://www.youtube.com/watch?v=jnoVtCKECmQ&t=611s) for more details.

Reviewer 2: a link to wiki is alr linked so just remove the vid.

Author: Alright... Also, the test worked earlier, but after changes, it failed :(...

<FocusProblem problem={problems.maxsum} />

### Solution

<Spoiler title="Solution">

> Consider the prefix sum array. The maximum subarray sum is the maximum possible $p[i]-p[j]$ where $j\leq i$. We can iterate over each index $i$ and keep a running minimum $p[j]$ value. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.

Reviewer 4: I think it would be helpful to elaborate on why this is.

Reviewer 5: also mention that $p$ is going to be the prefix sum array

Author: I commit @frodakcin 's change so i think i fixed...

> Consider the prefix sum array. The maximum subarray sum is the maximum possible $p[i]-p[j]$ where $j\leq i$. We can iterate over each index $i$ and keep a running minimum $p[j]$ value. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.

Reviewer 6: ```suggestion
Consider the prefix sum array $p$. The subarray sum over range $[i, j)$ where $j\leq i$ is $p[i]-p[j]$, so we are looking for the maximum possible value of $p[i]-p[j]$ over $0 \leq i \leq j \leq N$. 

For a fixed right bound $i$, the maximum subarray sum is

$$$
p[i]-\min_{j\leq i}{p[j]}
$$$

Thus, we can keep a running minimum to store $\min_{j\leq i}{p[j]}$ as we iterate through $i$. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.
```

Is this the elaboration you were looking for?

<LanguageSection>

<CPPSection>

C++ code:

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl dp(n+1, 0); // Prefix sum array initially filled with 0's
	for(int i=1; i<=n; i++){
		ll x; cin >> x;

> 		dp[i] = dp[i-1] + x; // Compute the prefix sum at each element

Reviewer 2: lol why call it dp when its psums. ig u could but that's confusing

Author: Lol i'm just used to writing it as dp cuz in one editorial: http://www.usaco.org/current/data/sol_paintbarn_silver_feb19.html, they used dp to represent prefix sums so im like y not lmao

	}
	ll max_subarray_sum = dp[0], min_prefix_sum = dp[0];
	for(int i=1; i<=n; i++){

> 			// Max subarray sum is the maximum difference between two prefix sums

Reviewer 6: I thought the standard was to have comments be on the same indentation as the code?

Reviewer 6: Looks fine tho

> 			// Max subarray sum is the maximum difference between two prefix sums

Reviewer 3: ```suggestion
		// Max subarray sum is the maximum difference between two prefix sums
```

		max_subarray_sum = max(max_subarray_sum, dp[i] - min_prefix_sum);
		min_prefix_sum = min(min_prefix_sum, dp[i]);
	}
	cout << max_subarray_sum << endl;
}
```

</CPPSection>

</LanguageSection>

</Spoiler>

<Warning>

This problem has a solution known as [Kadane's Algorithm](https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm). Please don't use that solution; try to solve it with prefix sums.

</Warning>

<Spoiler title = "Alternative Solution: Kadane's Algorithm">

Kadane's Algorithm finds the largest subarray sum using a greedy method. Check the video below for details:

> [Video explanation of the max subarray sum using Kadane's Algorithm](https://www.youtube.com/watch?v=jnoVtCKECmQ&t=611s).

Reviewer 2: nick white...

Reviewer 2: idk what are our opinions on putting nick white in guide LOL

Author: I mean he explains Kadane's fairly well lol... Also, not many ppl will look at Kadane's solution anyway lol..

Author: > nick white...

All that matters is that the approach is explained in a clear and logical way. Idt the channel really matters tbh. What matters is whether u could clearly learn, right??

Reviewer: I don't really like the explanation tbh, it seems geared towards an interview and not towards CP in general, so it's really long...

Author: If that's the case, then should we remove the Kadane's algorithm sol?

<LanguageSection>

<CPPSection>

> C++ code:

Reviewer 2: ```suggestion
## C++ Implementation
```

Reviewer 2: why was this resolved?

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl x(n);
	for(int i=0; i<n; i++){
		cin >> x[i];
	}
	ll current_sum = x[0], max_subarray_sum = x[0];
	for(int i=1; i<n; i++){
		current_sum = max(current_sum + x[i], x[i]); 

> 	// Continue the subarray sum or start a new subarray sum beginning at the current element

Reviewer 6: Yea fix this too

Reviewer 6: They  should be consistent at least

Author: ok.

<FocusProblem problem={problems.maxsum} />

### Solution

<Spoiler title="Solution">

> Suppose that we store the prefix sums at each index in the array. Intuitively, the maximum possible $p[i]-p[j]$ where $j\leq i$. Thus, we can iterate each element in the array and store the minimum prefix sum up to the current index. Then, we can find the maximum subarray sum by taking the difference of the current prefix sum and the minimum prefix sum.

Reviewer 6: ```suggestion
Consider the prefix sum array. The maximum subarray sum is the maximum possible $p[i]-p[j]$ where $j\leq i$. We can iterate over each index $i$ and keep a running minimum $p[j]$ value. This yields the maximum subarray sum for each possible right bound, and the maximum over all these values is our answer.
```

I think this is more clear. Maybe play around w the wording to make it more understandable.

<LanguageSection>

<CPPSection>

C++ code:

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl dp(n+1, 0); // Prefix sum array initially filled with 0's
	for(int i=1; i<=n; i++){
		ll x; cin >> x;
		dp[i] = dp[i-1] + x; // Compute the prefix sum at each element
	}
	ll max_subarray_sum = dp[0], min_prefix_sum = dp[0];
	for(int i=1; i<=n; i++){
	

> 	// Max subarray sum is the maximum difference between two prefix sums

Reviewer 6: What's up with the spacing here?

Author: Should I delete it?

<FocusProblem problem={problems.maxsum} />

### Solution - Max Subarray Sum

<Spoiler title="Solution">

Consider the prefix sum array $p$. The subarray sum $a_i \dots a_{j-1}$, where $i\leq j$ is $p[j]-p[i]$. Thus, we are looking for the maximum possible value of $p[j]-p[i]$ over $0 \leq i \leq j \leq N$. 

For a fixed right bound $j$, the maximum subarray sum is

> $$$
> p[j]-\min_{i\leq j}{p[i]}
> $$$

Reviewer 3: ```suggestion
$$
p[j]-\min_{i\leq j}{p[i]}
$$
```

<FocusProblem problem={problems.maxsum} />

> This problem has two types of solutions one of which is called the  [Kadane's Algorithm](https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm). However, understanding Kadane's algorithm is **optional** for this problem, since it can also be solved using prefix sums.

Reviewer 6: Original phrasing is better?

<Spoiler title="Solution">

> ### Solution

Author: Ok.

Suppose that we store the prefix sums at each index in the array. Intuitively, the maximum subarray sum ending at each index is the difference between the maximum prefix sum and the minimum prefix sum. Thus, we can iterate each element in the array and store the minimum prefix sum up to the current index. Then, we can find the maximum subarray sum by subtracting the current prefix sum to the minimum prefix sum.

> C++ code:

Reviewer 3: Remove or put this in the CPPSection

<CPPSection>

C++ code:

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl dp(n+1, 0); // Prefix sum array initially filled with 0's
	for(int i=1; i<=n; i++){
		ll x; cin >> x;
		dp[i] = dp[i-1] + x; // Compute the prefix sum at each element
	}
	ll max_subarray_sum = dp[0], min_prefix_sum = dp[0];
	for(int i=1; i<=n; i++){

> 		// The max subarray sum is the maximum difference between two prefix sums

Reviewer 3: ```suggestion
		// Max subarray sum is the maximum difference between two prefix sums
```
This is one character too long :(

		max_subarray_sum = max(max_subarray_sum, dp[i] - min_prefix_sum);
		min_prefix_sum = min(min_prefix_sum, dp[i]);
	}
	cout << max_subarray_sum << endl;
}
```

</CPPSection>

</LanguageSection>

</Spoiler>

<Warning>

Note: It is recommended that you skip the following solution.

</Warning>

<Spoiler title = "Alternative Solution">

### Solution

> Suppose we fix the first element of a subarray. The Kadane's Algorithm solution involves starting a new subarray sum at an element if that element is greater than the current subarray sum with an addition to the current element. If the subarray sum ending at the current element is greater than the current element, then we continue the subarray sum. Otherwise, we set the new subarray sum equal to the current element and restart our subarray sum. To find the maximum subarray sum over all possible sums, we store an integer $S$, representing the maximum subarray sum at each iteration of the array. When we finish iterating the array, we print $S$.

Reviewer 6: Not very fluid.

"if that element is greater than the current subarray sum with an addition to the current element" What does this mean?

"continue the subarray sum" what does this mean? Extend the current subarray to the right?

" To find the maximum subarray sum over all possible sums, we store an integer $S$, representing the maximum subarray sum at each iteration of the array. When we finish iterating the array, we print $S$." is the easy part. You can just simplify it to "print the largest subarray sum obtained by the algorithm" or something

Reviewer 6: Alternatively you can entirely waive the details by saying like "Kadane's Algorithm finds the largest subarray sum using a greedy method. Check the video below for details"

Watch this [video](https://www.youtube.com/watch?v=jnoVtCKECmQ&t=611s) for a more detailed explanation of the approach above.

> C++ code:

Reviewer 3: Same here - remove or put this in the CPPSection

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;

int main(){
	int n; cin >> n;
	vl x(n);
	for(int i=0; i<n; i++){
		cin >> x[i];
	}
	ll current_sum = x[0], max_subarray_sum = x[0];
	for(int i=1; i<n; i++){
		current_sum = max(current_sum + x[i], x[i]); 

> 	// Continue the subarray sum or start a new subarray sum beginning at the current element

Reviewer 3: shorten or put this on two lines (also fix the indentation)

Reviewer 3: Yep

Author: Should I remove the tab or should I add another tab??

Reviewer 3: Add another tab
